### DISCUSSION TITLE
Dependabot Guidance

### DISCUSSION THREAD

**Discussion by mvs5465** [projen/projen#2460] [Q&A]
Hi all, I opened this feature-request ticket on Dependabot: https://github.com/dependabot/dependabot-core/issues/6708

It appears that currently Projen supports Dependabot, but not the other way around. Since Dependabot does not update packages in `.projenrc.js` it appears to be fundamentally incompatible right now.

For example if we set the `aws-cdk` version in our `.projenrc.js` file and Dependabot attempts to update it in package.json, the next time a user runs `npx projen` the updates to package.json will be reverted and overwritten.

Is this valid thinking? Is anybody here successfully using both projen and dependabot?

**Comment by mrgrain** [comment]
Thanks for initiating this discussion, it's an important one! Please note I'm not a user of Dependabot, so my experience is limited. Even more so for projects that are not based in the NPM ecosystem (i.e. Python, Java). But here are my initial thoughts on this topic from a perspective of how projen works.

---

projen manages dependencies in code. **The mechanism to delegate version updates to the outside of `.projenrc.js` is by defining the dependency without a version constraint.** You can observe this feature by looking at `.projen/deps.json`. Only dependencies with explicit constraints will have them listed there.

**When a constraint is set**, the assumption is that a developer made a conscious decision to constraint the dependency to a specific version. And **updating automatically beyond the constraint is likely not desired at all**. This might be the case for dependencies directly defined by the end-user, in which case the end-user can adapt their behavior. But constraints might also come from an upstream project (~= transitive dependency). We may need to be better at documenting or highlighting this mechanism. For upstream dependencies, I'm not sure what the answer would be. E.g. npm already does have a mechanism for this (`overrides`), but I don't think Dependabot actually supports it. **Projen may well need something similar to overrides**. (Although as a side note, for node they can already define package resolutions in code). I also do not know if this 

In my mind, we've now established that **projen dependency constraints should be respected.** What needs to happen next is that we MUST tell all dependency update mechanism to respect them. For the "projen-native" upgrades via GH Action Workflows, we definitely support this. If we don't yet for Dependabot (and also Renovate) we should add this!
